pennylane/gradients/pulse_generator_gradient.py

```diff
@@ -234,3 +250,210 @@ def _generate_tapes_and_coeffs(tape, idx, atol, cache):
     # Return the start and end pointer for the created tapes/the corresponding results.
     # Also return the coefficients for the current trainable parameter of the pulse
     return tapes, (start, end, all_coeffs[term_idx]), cache
+
+
+def _parshift_and_contract(results, coeffs, single_measure, shot_vector):
+    """Compute parameter-shift tape derivatives and contract them with coefficients.
+
+    Args:
+        results (list[tensor_like]): Tape execution results to be processed.
+        coeffs (list[tensor_like]): Coefficients to be contracted.
+        single_measure (bool): whether the tape execution results contain single measurements.
+        shot_vector (bool): whether the tape execution results were obtained with a shot vector.
+
+    Returns:
+        tensor_like or tuple[tensor_like] or tuple[tuple[tensor_like]]: contraction between the
+        parameter-shift derivative computed from ``results`` and the ``coeffs``.
+    """
+
+    def _parshift_and_contract_single(res_list, coeffs):
+        """Execute the standard parameter-shift rule on a list of results
+        and contract with Pauli basis coefficients."""
+        psr_deriv = ((res := qml.math.stack(res_list))[::2] - res[1::2]) / 2
+        return qml.math.tensordot(psr_deriv, coeffs, axes=[[0], [0]])
+
+    if single_measure and not shot_vector:
+        # single measurement and single shot entry
+        return _parshift_and_contract_single(results, qml.math.stack(coeffs))
+    if single_measure or not shot_vector:
+        # single measurement or single shot entry, but not both
+        return tuple(
+            _parshift_and_contract_single(r, qml.math.stack(coeffs)) for r in zip(*results)
+        )
+
+    return tuple(
+        tuple(_parshift_and_contract_single(_r, qml.math.stack(coeffs)) for _r in zip(*r))
+        for r in zip(*results)
+    )
+
+
+def _expval_pulse_generator(tape, argnum, shots, atol):
+    """Compute the pulse generator parameter-shift rule for a quantum circuit that returns expectation
+    values of observables.
+
+    Args:
+        tape (`~.QuantumTape`): Quantum circuit to be differentiated with the pulse generator
```

Comment:
Or maybe not, if it currently can't be?

Reply:
The "If the input is a QNode," blurb below kind of throws me then, because it makes it sound like an option - but I guess this is how the other ones do it too 😅